File: tutorial3-complete/src-ios/Libraries/FoyerKit/Sources/HomeShell/HomeBuilder.swift
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

import FoyerBridge
import FoyerKit
import Foundation

/// What the home tab consumes from its parent: the items port.
/// sourcery: DuetComponent
/// sourcery: CreateMock
public protocol HomeDependency: AnyObject {
  var items: any ItemsPort { get }
}

final class LiveHomeEnvironment: NSObject, HomeEnvironment {
  private let items: any ItemsPort

  init(items: any ItemsPort) {
    self.items = items
  }

  func loadItems(onItems: @escaping ([Item]) -> Void) {
    items.items(onItems: onItems)
  }
}

extension HomeComponent {
  @MainActor
  func environment() -> LiveHomeEnvironment {
    LiveHomeEnvironment(items: items)
  }
}

// MARK: - Builder

public final class HomeChild {
  public let shell: HomeViewShell

  init(shell: HomeViewShell) {
    self.shell = shell
  }
}

public final class HomeBuilder {
  private let dependency: HomeDependency

  public init(dependency: HomeDependency) {
    self.dependency = dependency
  }

  @MainActor
  public func buildHome() -> HomeChild {
    let component = HomeComponent(dependency: dependency)
    let scope = mainImmediateStoreScope()
    let store = makeHomeStore(environment: component.environment(), scope: scope)
    let bridged = HomeKitStore(
      state: homeStateFlow(store: store),
      send: { store.send(action: $0) },
      teardown: {
        store.teardown()
        cancelStoreScope(scope: scope)
      })
    return HomeChild(shell: HomeViewShell(store: bridged))
  }
}

File: tutorial4-start/src-kmp/app/src/main/kotlin/dev/modaal/foyer/app/HomeBuilder.kt
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

package dev.modaal.foyer.app

import dev.modaal.duet.kernel.Store
import dev.modaal.foyer.home.HomeAction
import dev.modaal.foyer.home.HomeEffectPayload
import dev.modaal.foyer.home.HomeEnvironment
import dev.modaal.foyer.home.HomeState
import dev.modaal.foyer.home.makeHomeStore
import dev.modaal.foyer.ports.Item
import dev.modaal.foyer.ports.ItemsPort
import kotlinx.coroutines.CoroutineScope

typealias HomeStore = Store<HomeState, HomeAction, HomeEffectPayload>

/** What the home tab consumes from its parent: the items port. */
interface HomeDependency {
  val items: ItemsPort
}

class HomeComponent(dependency: HomeDependency) : HomeDependency by dependency {
  fun environment(): HomeEnvironment =
    object : HomeEnvironment {
      override fun loadItems(onItems: (List<Item>) -> Unit) = items.items(onItems)
    }
}

class HomeBuilder(private val dependency: HomeDependency) {
  fun buildHome(scope: CoroutineScope): HomeStore {
    val component = HomeComponent(dependency)
    return makeHomeStore(environment = component.environment(), scope = scope)
  }
}
